### ISSUE TITLE
Infer type of `attrs.fields(type(attrs_instance))`

### ISSUE THREAD

**Issue by injust** [python/mypy#17426]
**Feature**

mypy currently infers the type of `attrs.fields(foo)` as `Any`, where `foo` is an instance of an `attrs` class. Is it possible to have mypy infer the correct type?

With this setup:
```python
import attrs
from attrs import define


@define
class Foo:
    bar: int


foo = Foo(1)
```

mypy infers the correct type for `attrs.fields(Foo)`:
```python
fields = attrs.fields(Foo)
reveal_type(fields)
reveal_type(fields.bar)
reveal_type(fields.not_bar)

"""
demo.py:13: note: Revealed type is "tuple[attr.Attribute[builtins.int], fallback=demo.Foo.__demo_Foo_AttrsAttributes__]"
demo.py:14: note: Revealed type is "attr.Attribute[builtins.int]"
demo.py:15: error: "__demo_Foo_AttrsAttributes__" has no attribute "not_bar"  [attr-defined]
demo.py:15: note: Revealed type is "Any"
Found 1 error in 1 file (checked 1 source file)
"""
```

but mypy infers `attrs.fields(type(foo))` as `Any`:
```python
fields = attrs.fields(type(foo))
reveal_type(fields)
reveal_type(fields.bar)
reveal_type(fields.not_bar)

"""
demo.py:13: note: Revealed type is "Any"
demo.py:14: note: Revealed type is "Any"
demo.py:15: note: Revealed type is "Any"
Success: no issues found in 1 source file
"""
```

**Pitch**

Passing the class to `attrs.fields()` is the [documented](https://www.attrs.org/en/stable/api.html#attrs.fields) usage.

If you pass `type(attrs_instance)` instead, mypy quietly treats the type as `Any` (even under `strict` mode), so it cannot catch `attr-defined` errors when you access a non-existent field.

ref https://github.com/python-attrs/attrs/issues/1297

**Comment by injust** [comment]
@Tinche called out [here](https://github.com/python-attrs/attrs/issues/1297#issuecomment-2184132255):
> There's probably a difference between `Foo` and `type(foo)` - the latter can be any arbitrary subclass of `Foo`.

But even if you use `attrs.fields()` in a method of an `attrs` class, `mypy` still infers `attrs.fields(type(self))` and `attrs.fields(cls)` as `Any`.
